### PR TITLE
M1a: authentication and web-surface security

### PR DESCRIPTION
**Pull request by bassings**
## M1a: authentication and web-surface security

Implements PR 2 of `specs/REMEDIATION-2026-08.md`. Every change is test-first,
and every fix was proven load-bearing by breaking the thing it guards, watching
the right test fail, and restoring the file byte-identically by SHA-256.

**Four of these are live authentication bypasses, and they compound.**

---

### The bypasses

**Auth was gated on `if username and password:`** (`__init__.py:57-71`), and
returned "fully public" otherwise. Driven directly across all four
combinations:

```
username='admin' password='secret'  -> denied without a cookie
username=''      password='secret'  -> PUBLIC
username='admin' password=''        -> PUBLIC
username=''      password=''        -> PUBLIC
```

So an operator who set a password but left the username blank had a wide-open
server and nothing said so. That was not an unlikely mistake: the settings copy
put **"Leave empty to disable authentication" on the USERNAME field**, so the
field that silently disabled auth was the one whose own description promised it
would. Now gated on an explicit `auth_required` setting; the copy has moved.

**`/login/` issued a valid session for any credentials when no password was
set.** `... or not password` short-circuits the whole credential check to
`True`. The cookie it writes is the **api_key**, so this is not a UI-only
bypass — it hands the caller the credential the entire API authenticates with.

**`/getkey/` had the same flaw and is worse in consequence.** It returns the
api_key itself, in a JSON body, over GET, with credentials in the query string
— so the request line lands in access logs, proxy logs and browser history. The
`u` parameter is no protection: it is the md5 of the username, which is not a
secret.

**Passwords set through the UI or wizard were stored as unsalted MD5.**
`setting.save.core.password` was wired to `md5(value)`. bcrypt was reached only
by the login-time upgrade, which never runs for anyone who has not logged in —
and until `auth_required` landed here, the server never asked anyone to log in.
So the cohort this protects is exactly the cohort that was still on MD5.

`check_password` carried the comment **"New passwords are always bcrypt"**,
false from the day it was written. Corrected, with a test pinning it — that
sentence is *why* the defect survived, because it told every reader the thing
in front of them had already been handled.

### Why they compound

Until T2.0 landed in this PR, `config.ini` was created **0644 — world-readable**
— holding the api_key, that MD5 password, and every downloader and notifier
credential. The endpoint that leaked the key logged its own credentials in the
query string. And the log filter only redacted query-string secrets, so a bot
token logged at ERROR went out in the clear.

---

### Data safety

**`Settings.save()` was truncate-then-write.** `open(self.file, 'w')` truncates
before a byte is written, so a process death or full volume mid-write left
`config.ini` empty or partial. That file is also the documented lock-out
recovery path, so the failure destroyed the remedy along with the
configuration. Now: sibling temp file, fsync, `os.replace`, fsync the
directory.

Four details the obvious fix gets wrong, each checked against the tree rather
than assumed: `realpath` first (else `os.replace` replaces a symlinked config
and orphans the real file), a **sibling** temp (production bind-mounts
`/config`, so `/tmp` would break the rename), `BaseException` on cleanup (a
Ctrl-C must not leave `.config.ini.*.tmp`, which is what a locked-out operator
reaches for), and fsync of the directory.

### The class fix

**Route auth inventory** — every route must either carry the auth dependency or
sit in `PUBLIC_ROUTES` with its reason written inline. `/getkey/` exists
because that decision was never written down anywhere; this closes the class
rather than the instance.

It walks `_IncludedRouter.original_router` recursively, and that detail is
load-bearing: **FastAPI 0.140 stopped flattening included routers**, so a naive
`for r in app.routes` sees 14 routes and **zero** with `require_auth`, against a
real 80 routes / 33 protected paths. A broken walk does not merely undercount,
it *inverts* the answer. I produced that exact false alarm twice while writing
the file, which is why it opens with an anti-vacuity test.

---

### Deliberately not done, and why

- **HMAC-signed session cookie.** The cookie is still the api_key. This is the
  right fix and is its own PR: it changes how every request authenticates, and
  it signs out every logged-in user, so it wants its own before/after on a real
  instance.
- **`/getkey/` deleted.** Its only referrer is `simple_healthcheck.py:76`,
  whose removal is blocked on AC-OPS-12's production grep. Closing the hole did
  not depend on that, so the hole is closed and the deletion follows.
- **T2.6's Synology TLS item — RETRACTED in the spec, with evidence.**
  `synology.py:109-110` hardcodes `http://`, so the `verify=False` it targets is
  unreachable and "honour ssl_verify" would have shipped a commit that reads
  like TLS hardening and hardens nothing. The real exposure is that `_login`
  POSTs `account`/`passwd` over plaintext HTTP and the `host` setting accepts no
  scheme. That changes downloader transport and cannot be exercised without a
  real Synology, so it is deferred rather than guessed at.

### Corrections made in this branch

Stated rather than tidied away:

1. A `/getkey/` leak test **passed against the unfixed code** on its first run —
   it sent `u=anything`, so the username clause rejected the request before the
   password clause was reached. It looked like a leak test while measuring the
   wrong half of the condition. Rewritten.
2. The route-inventory test was **deleted once and rewritten**, because the
   first version introspected an app it did not understand and reported "0
   routes protected".
3. `auth_required` is a plain `{'default': 0}`, not the tri-state the spec first
   sketched — `registerDefaults` materialises the literal `auth_required = None`
   into config.ini and `Env.setting`'s default is `''`, so the natural read is
   falsy and auth would stay off on every install with no failing test and no
   log line.

### Dependencies

Triaged before raising this PR, per the rule added to the global standards this
session. Eight open Dependabot PRs; **six taken, two rejected**, and the
rejections are the point:

| PR | Outcome |
|---|---|
| #219 fastapi 0.141.1 | **taken** — verified first by unpacking the wheel, because this PR's route inventory walks `_IncludedRouter`, introduced in 0.140. Still present in 0.141.1. |
| #222 ruff 0.16.1 | **taken as an EXACT pin**, not Dependabot's `>=0.16.1`. The floating form is what let `make lint` and the gate disagree about the version earlier. Both ci.yml jobs bumped with it — a test caught me having done only one. |
| #220 jsdom, #218 playwright, #217 brace-expansion | **taken**. #217 is transitive: a plain install would not move it, it needed `npm update`. |
| **#224 trakit 0.3.0** | **REJECTED.** `pip check` after applying: `knowit 0.5.11 has requirement trakit<0.3.0`. 0.5.11 is the latest release and every version back to 0.5.8 carries the same cap, so no combination satisfies both. Merging it dutifully would have left an environment `pip check` rejects. |
| **#221 pyOpenSSL, #223 cryptography** | **Already done** in PR 1, as the fix for CVE-2026-69247. Redundant; close them. |

Separately, `npm audit` reported six advisories not covered by any Dependabot
PR — 4 high, 1 moderate, 1 low, all transitive dev dependencies. Now **zero**,
via `npm audit fix` without `--force`, so `package.json` is unchanged and only
resolved transitive versions moved.

Both graphs verified after applying: `npm ls` resolves, `pip check` reports no
broken requirements.

### Verification

- `make verify` green end to end from a clean state.
- **2291 Python unit + 42 integration + 202 UI unit**; 135 chromium, 2
  isolation (2 workers), 2 mobile, 21 accessibility.
- Every fix reverted and re-run: the named tests red, the file restored
  byte-identical.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Ss6TA5seR8ykyJfe3itaSc

### CODE REVIEW

File: couchpotato/core/_base/_core.py [hunk 1]
```diff
@@ -87,7 +87,33 @@ def dependencies(self):
...
+        every existing and future user. Changing the inner encoding needs its
+        own migration and login-time upgrade, and is not a tidy-up to fold in
+        here.
+
+        The event name (`setting.save.core.password`) and this method name stay
+        as they are: renaming them is a separate change with its own blast
+        radius, and a wrong name with a correct docstring is safer than a
+        rename made at the same time as a security fix.
+        """
+        return hash_password(md5(value)) if value else ''
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/_base/_core.py:165]: Confirmed and fixed — and you were more right than the finding says. `setting.save.core.password` now maintains `auth_required` in **both** directions, and the copy now matches.

The second direction is not symmetry for its own sake. `auth_required = 1` with the password cleared meant `get_current_user` denied every request **and** `login_post` refused every credential (it now requires a configured password, from the blank-password fix earlier in this PR) — a total lockout recoverable only by hand-editing config.ini, reachable by setting a password and later clearing it. Neither of those two fixes caused it alone; together they did. I found that while fixing yours.

The write is guarded so it cannot break password hashing, but logged at ERROR naming the consequence rather than swallowed — a silent failure here restores the exact state this closes.

File: couchpotato/core/settings.py [hunk 2]
```diff
@@ -234,8 +236,83 @@ def getValues(self):
...
+            prefix='.%s.' % os.path.basename(path), suffix='.tmp', dir=directory
+        )
+        try:
+            with os.fdopen(tmp_fd, 'w', encoding='utf-8') as configfile:
+                self.p.write(configfile)
+                configfile.flush()
+                os.fsync(configfile.fileno())
+
+            os.chmod(tmp_path, mode)
+            os.replace(tmp_path, path)
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/settings.py:339]: Both settings.py findings confirmed and fixed, and they share a root cause worth naming: the truncate-then-write this replaced preserved the **inode**, so owner, group and ACLs survived for free. `os.replace` creates a new one, so they have to be copied deliberately.

- **Ownership**: the existing gid is now carried across, best-effort — chown to a different uid needs privilege this process will not have, and a failure must not fail the save. ACLs beyond the mode bits are explicitly **not** preserved and that is documented rather than half-attempted.
- **World bits**: now always cleared. You are right that mode-preservation meant every pre-existing 0644 config stayed world-readable forever — the fix hardened new installs and did nothing for the ones that already had the problem. Only the world bits; group access is left as the operator set it, since backup tooling on a shared group is legitimate.

Worth flagging: my first ownership test passed trivially, because same-uid meant `mkstemp`'s default reproduced the expected value by accident. Rewritten to chgrp to a non-primary group first, which `mkstemp` cannot reproduce — it then failed as it should.

File: couchpotato/core/_base/_core.py [hunk 3]
```diff
@@ -269,19 +295,36 @@ def signal_handler(*args, **kwargs):
...
             'wizard': True,
             'options': [
+                {
+                    'name': 'auth_required',
+                    'default': 0,
+                    'type': 'bool',
+                    'label': 'Require login',
+                    'description': 'Require login for the web interface. Turned on '
+                                   'automatically the first time a password is set. '
+                                   'Turn off only for a trusted LAN.',
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/_base/_core.py]: Confirmed and fixed — and you were more right than the finding says. `setting.save.core.password` now maintains `auth_required` in **both** directions, and the copy now matches.

The second direction is not symmetry for its own sake. `auth_required = 1` with the password cleared meant `get_current_user` denied every request **and** `login_post` refused every credential (it now requires a configured password, from the blank-password fix earlier in this PR) — a total lockout recoverable only by hand-editing config.ini, reachable by setting a password and later clearing it. Neither of those two fixes caused it alone; together they did. I found that while fixing yours.

The write is guarded so it cannot break password hashing, but logged at ERROR naming the consequence rather than swallowed — a silent failure here restores the exact state this closes.

File: couchpotato/core/settings.py [hunk 4]
```diff
@@ -234,8 +236,83 @@ def getValues(self):
...
+        """
+        path = os.path.realpath(self.file)
+        directory = os.path.dirname(path) or '.'
+
+        try:
+            # `& 0o777` rather than stat.S_IMODE: `stat` is not imported here,
+            # and this fix should not need to add a module-level import.
+            mode = os.stat(path).st_mode & 0o777
+        except OSError:
+            mode = 0o600
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/settings.py:289]: Both settings.py findings confirmed and fixed, and they share a root cause worth naming: the truncate-then-write this replaced preserved the **inode**, so owner, group and ACLs survived for free. `os.replace` creates a new one, so they have to be copied deliberately.

- **Ownership**: the existing gid is now carried across, best-effort — chown to a different uid needs privilege this process will not have, and a failure must not fail the save. ACLs beyond the mode bits are explicitly **not** preserved and that is documented rather than half-attempted.
- **World bits**: now always cleared. You are right that mode-preservation meant every pre-existing 0644 config stayed world-readable forever — the fix hardened new installs and did nothing for the ones that already had the problem. Only the world bits; group access is left as the operator set it, since backup tooling on a shared group is legitimate.

Worth flagging: my first ownership test passed trivially, because same-uid meant `mkstemp`'s default reproduced the expected value by accident. Rewritten to chgrp to a non-primary group first, which `mkstemp` cannot reproduce — it then failed as it should.

File: couchpotato/core/logger.py [hunk 5]
```diff
@@ -79,6 +98,18 @@ def filter(self, record):
...
+        #
+        # Stops at whitespace or at any of `&,;)]}'"` so the REST of the
+        # message survives -- `token=SECRET and the response was 404` must keep
+        # the 404. A filter that swallows the line removes the reason the line
+        # was logged, and the operator turns it off.
+        #
+        # \b on the left so `mytoken=` is left alone rather than
+        # half-redacted.
+        for replace in _REPLACE_PRIVATE_BARE:
+            msg = re.sub(r'\b(%s=)[^\s&,;)\]}\'"]+' % replace, r'\1xxx', msg)
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/logger.py]: Not a nit — a live leak, and thank you for the pointer to `plex/server.py:85`. That URL goes to `urlopen`, and `http_client.py:236` logs the whole URL at INFO, so the Plex auth token was written to the log on every notification. Neither pass caught it, exactly as you describe.

Bare matching is now case-insensitive with an optional `[\w-]*` prefix, so `access_token=`, `bot_token=` and `X-Plex-Token=` are all redacted. `sid` is kept exact-match — prefix-matching three characters eats ordinary words, and I added a test (`test_short_names_are_not_prefix_matched`) to hold that line, since a redaction people cannot read around is one they switch off.

File: couchpotato/runner.py [hunk 6]
```diff
@@ -269,6 +269,29 @@ def customwarn(message, category, filename, lineno, file=None, line=None):
...
+        Env.setting('auth_required', value=resolved)
+        log.info('auth_required was not set; resolved to %s from the stored '
+                 'password and written to the settings file', resolved)
+
+    if not auth_is_required():
+        # WARNING, not INFO: an unauthenticated instance is a decision, and the
+        # log is where an operator checks whether they made it.
+        log.warning('Serving with authentication DISABLED -- anyone who can '
+                    'reach this port has full control of the library. Set a '
+                    'password and turn on "Require login" in Settings.')
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/runner.py:293]: Confirmed and pinned. You're right that nothing enforced it: move that block below `loader.run()` and `registerDefaults` → `setDefault` will have materialised `auth_required = 0`, so the "absent" check skips and every upgraded install that had a password becomes public — no failing test, no log line, and a diff that looks like tidying.

Added a source-order assertion in `test_auth_required_gate.py`. A blunt instrument, and I considered making the migration robust to ordering instead, but every version of that still has to distinguish "never set" from "set to the registered default", which config.ini cannot express once written. So the ordering *is* the mechanism, and pinning it beats documenting it — this branch has already learned what a comment is worth once the code stops matching it.

File: couchpotato/core/_base/_core.py [hunk 7]
```diff
@@ -87,7 +87,70 @@ def dependencies(self):
...
+        # theirs to set. This only moves it when the password itself changes.
+        # Guarded, and NOT silently: hashing the password must not fail
+        # because a secondary write did, but a swallowed failure here puts the
+        # instance back in the state this whole block exists to prevent -- so
+        # it is logged at ERROR, naming the consequence, rather than passed
+        # over. (It also keeps this hook callable in isolation, e.g. from a
+        # unit test, where no settings store is loaded.)
+        from couchpotato.environment import Env
+        try:
+            Env.setting('auth_required', value=1 if value else 0)
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/_base/_core.py]: Confirmed and fixed — you're right that it reintroduced the lockout through a different door, and the mechanism is exactly as you describe: `Env.setting(value=...)` saves on the spot, then `saveView` saves again afterwards, so `auth_required = 1` reached disk before the password. A crash in between leaves authentication on, no password stored, and no way back short of hand-editing config.ini.

The hook now touches only the in-memory parser and lets `saveView`'s single `save()` persist both. That save is atomic (T2.0), so the pair lands together or not at all.

Pinned by asserting the **mechanism** rather than the outcome: once both saves complete, nothing observable distinguishes one from two — the difference only exists if the process dies in between. So one test asserts the hook triggers no save of its own, and a second asserts the value still reaches the settings object, because not-saving must not quietly become not-setting.

File: couchpotato/core/_base/_core.py [hunk 8]
```diff
@@ -87,7 +87,70 @@ def dependencies(self):
...
+        # An operator who wants a password AND an open instance can still turn
+        # "Require login" off explicitly afterwards -- that setting remains
+        # theirs to set. This only moves it when the password itself changes.
+        # Guarded, and NOT silently: hashing the password must not fail
+        # because a secondary write did, but a swallowed failure here puts the
+        # instance back in the state this whole block exists to prevent -- so
+        # it is logged at ERROR, naming the consequence, rather than passed
+        # over. (It also keeps this hook callable in isolation, e.g. from a
+        # unit test, where no settings store is loaded.)
+        from couchpotato.environment import Env
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/_base/_core.py]: Taken — right on both counts. The local import is gone, and the docstring's claim that it kept the hook callable in isolation was simply wrong: the module-scope import at line 14 already did that. Removing the sentence mattered as much as removing the import, since it was a reason that had never been true.

File: couchpotato/__init__.py [hunk 9]
```diff
@@ -54,22 +54,53 @@ def get_db():
...
+    `Env.setting('auth_required', type='bool')` reads falsy and auth stays off
+    on every install, silently, with no failing test and no log line.
+    """
+    configured = Env.setting('auth_required', default=None)
+
+    if configured is None or configured == '':
+        return bool(Env.setting('password'))
+
+    if isinstance(configured, str):
+        return configured.strip().lower() in ('1', 'true', 'yes', 'on')
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/__init__.py]: Not a nit — writing the missing test found a live defect, fixed in 1591536b.

You are right on every step of the mechanism, and I verified each: `runner.py:288` precedes `loader.run()` at `:333`; `getType` falls back to `'unicode'` (`settings.py:375-379`); `'unicode'` is not a key in `_type_adapters` (`settings.py:16-21`), so `_coerce_value` returns the raw ConfigParser string. The startup path is the *string* path, not an edge case.

What the test exposed once written: the branch returned `False` for anything outside its truthy set. So `auth_required = yess` — a typo — read as "off" and made a password-protected server **public**. `config.ini` is the documented lock-out recovery path, so it gets hand-edited and typos there are expected.

Neither fixed answer works. Reading garbage as "off" is that bug; reading it as "on" turns auth on for an install with no password, which nothing can satisfy — the lockout this branch has already had to close twice. An unrecognised value now derives from whether a password is set, which errs shut exactly when there is a credential to open it with, and logs an error naming the bad value.

Proven load-bearing with three mutations, file restored byte-identical by SHA-256 each time:
- remove the `str` branch entirely → 7 fail, every falsy-string case
- unrecognised returns `False` (the original defect) → 1 fail, the right one
- unrecognised returns `True` (lockout) → 1 fail, the other direction

File: couchpotato/core/settings.py [hunk 10]
```diff
@@ -234,8 +236,128 @@ def getValues(self):
...
+          not exist yet IS created 0600 deliberately: it holds credentials,
+          and 0644 -- what the old code produced under a default umask -- made
+          every secret in it world-readable.
+
+        - fsync the file before the rename and the directory after it.
+          Otherwise a power failure can leave the rename durable while the
+          contents are not, which is the same lost-config outcome by a
+          slower route.
+        """
+        path = os.path.realpath(self.file)
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/settings.py:274]: Agreed it is not a regression, and I want to record why `realpath` is the deliberate choice rather than the tolerated one — because the alternative is worse, not merely different.

Dropping `realpath` does not stop the symlink being followed; it changes `os.replace()` to replace **the symlink itself** with a regular file. The operator’s real config is then orphaned: their deliberate symlink silently stops working, subsequent saves go to a different file from the one they are reading, and the original is left behind holding stale credentials. That is a data-integrity regression traded for no security gain.

On the threat model: `config.ini`’s path comes from `--config_file` (`runner.py:57-58,80-86`), a command-line argument fixed at process start. There is no request-reachable path into it — I checked; nothing outside argument parsing calls `setFile`. So the actor who controls the path is the actor who starts the process, which is the same actor who controls the file. Following the symlink grants them nothing they did not already have.

Your conditional is the right one to flag, though, and it is the part worth keeping: if the path ever becomes reachable from a request or a less-trusted config surface, this call is where it bites, and the fix then is `O_NOFOLLOW` on the target plus refusing to write through a symlink whose owner differs from the process. I have not built that now because it would be a guard against a caller that does not exist, and CLAUDE.md §9 asks that a rule be enforced only where an agent actually fails without it.

Resolving as verified-and-deliberate rather than actioned.

File: couchpotato/core/_base/_core.py [hunk 11]
```diff
@@ -269,19 +344,37 @@ def signal_handler(*args, **kwargs):
             'description': 'Needs restart before changes take effect.',
             'wizard': True,
             'options': [
+                {
+                    'name': 'auth_required',
+                    'default': 0,
+                    'type': 'bool',
+                    'label': 'Require login',
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/_base/_core.py:351]: Fair, and I checked the "presumably" rather than taking it — that assumption is the load-bearing part, and this repo has already shipped a settings block that was faithful to the existing pattern and completely unreachable (`fanarttv.py` copied `themoviedb`'s `tab: 'providers'` + `hidden: True`, which `hiddenTabs` filters out entirely, so the key could only be set by hand-editing config.ini).

It does render, and here is why:

- `partials/settings/scripts.html:133` — `hiddenTabs: new Set(['providers', 'automation'])`. `auth_required` is on `tab: 'general'`, which is not filtered.
- `partials/settings/field_types.html:5` — `x-if="opt.type === 'bool'"` renders the checkbox; `auth_required` is `type: 'bool'`.
- `settings.py:405` — `getOptions` gates on `isOptionReadable`, which only excludes an option carrying a `_internal_meta` value outside `('rw','ro')`. `auth_required` has no `ui-meta` at all, so it passes.
- By the time the settings API is called, `registerDefaults` has run `setType`, so `Settings.get` coerces the stored `0`/`1` to a real bool for the checked state — the string path that matters at startup (and that this PR's last commit fixed) does not apply here.

So the field is visible, labelled "Require login", and reflects the persisted value.

**Taking the E2E assertion as a follow-up, not doing it here.** It is coverage rather than a defect, you have marked it non-blocking, and adding it costs a full ~35-minute local gate on a PR that is otherwise green with every check passing. I am running a gate on the next branch shortly, so it goes there rather than being dropped — recorded so it is not lost.

The narrower point stands and I have not addressed it: nothing currently fails if that checkbox stops rendering. That is a real gap in a security-relevant control, which is exactly why it is being tracked rather than closed.
